pkg/queue/handlers/http_request.go

```diff
@@ -0,0 +1,258 @@
+	client          *http.Client
+}
+
+func (h *httpRequestHandler) Process(ctx context.Context, task queue.Task, heartbeats chan<- queue.Progress) (err error) {
+	span, ctx := h.StartSpan(ctx, "Process")
+	defer func() {
+		close(heartbeats)
+		h.FinishSpan(span, err)
+	}()
+	span.SetTag("task.id", task.ID)
+	span.SetTag("task.queue", task.Queue)
+	span.SetTag("task.type", task.Type)
+	span.SetTag("task.spec", string(task.Spec))
+
+	logrus := logrus.WithField("type", task.Type).WithField("id", task.ID)
+
+	logrus.Debug("starting the HTTP request task...")
+
+	var progress HTTPRequestProgress
+	defer func() {
+		// we check for errSerializingHearbeat so we don't cause
+		// a recursion call
+		if err == nil || err == ErrSerializingHearbeat {
+			return
+		}
+		message := err.Error()
+		progress.ErrorMessage = &message
+		_ = sendHTTPRequestProgress(progress, heartbeats)
+	}()
+
+	var spec HTTPRequestTaskSpec
+	err = json.Unmarshal(task.Spec, &spec)
+	if err != nil {
+		return err
+	}
+
+	progress.Stage = RequestPreparing
+	err = sendHTTPRequestProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	var payload io.Reader
+	if spec.RequestBody != "" {
+		payload = strings.NewReader(spec.RequestBody)
+	}
+
+	req, err := http.NewRequest(spec.Method, spec.URL, payload)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("User-Agent", "Contiamo Hub HTTP Request Task")
+
+	for name, value := range spec.RequestHeaders {
+		req.Header.Add(name, value)
+	}
+
+	if spec.Authorized {
+		token, err := h.tokenCreator.Create("httpRequestTask")
+		if err != nil {
+			return err
+		}
+
+		req.Header.Add(h.tokenHeaderName, token)
+	}
+
+	err = opentracing.GlobalTracer().Inject(
+		span.Context(),
+		opentracing.HTTPHeaders,
+		opentracing.HTTPHeadersCarrier(req.Header),
+	)
+	if err != nil {
+		otext.Error.Set(span, true)
+		span.SetTag("tracing.inject.err", err.Error())
+		err = nil
+	}
+
+	progress.Stage = RequestPending
+	err = sendHTTPRequestProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	defer func() {
+		duration := time.Since(now).Milliseconds()
+		progress.Duration = &duration
+		err := sendHTTPRequestProgress(progress, heartbeats)
+		if err != nil {
+			logrus.Error(err)
+		}
+	}()
+
+	resp, err := h.client.Do(req.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	contentType := resp.Header.Get("content-type")
+	if !strings.Contains(contentType, "application/json") {
+		return fmt.Errorf(
+			"unexpected response content type, expected `application/json`, got `%s`",
+			contentType,
+		)
+	}
+
+	progress.Stage = RequestResponse
+	progress.ReturnedStatus = &resp.StatusCode
+	err = sendHTTPRequestProgress(progress, heartbeats)
+	if err != nil {
+		return err
+	}
+
+	// the task would time out if the heartbeat was not sent for 30 seconds
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	go func() {
+		for {
+			select {
```

Comment:
Not having this select and ranging over the `ticket.C` channel caused the go routine leak detection. :man_shrugging:

Reply:
Strange. I mean you stop the ticker explicitly oO